### ISSUE TITLE
Is there any tools that do post fine tuned mesh after completed MeshSegNet Step 6.

### ISSUE THREAD

**Issue by ly000451** [Tai-Hsien/MeshSegNet#8]
I have an annoying issue.  Generally, the segmented mesh is not so accurate and the boundary is not very smooth. I need do post fine-tuned work.  Meshlabeler can do initial mesh training work but it is challenging to do the post fine-tuned work. so I want to know Is there any other tools to do post fine-tuned label mesh work. Thanks so much.

**Comment by Tai-Hsien** [comment]
Hello @ly000451 
When you say "segmented mesh", do you mean the result from MeshSegNet with graph-cut (i.e., step6)? or without graph-cut (i.e., step5)? The graph-cut algorithm is for post fine tune, although it's still not perfect. Some people also use conditional random field (CRF) to achieve the post fine tune. However, to implement these fine tune processes, I usually do it with coding and haven't see any other GUI tool for it. Maybe there are some, but I just don't know.

Feel free to let me know if I misunderstand what you mean.

**Comment by ly000451** [comment]
yes，you got it right. after I do complete step 6, I find the result of segmented tooth crown mesh boudnary is not accurate. so I want to do some fine-tuned to the step 6 result manually.  if that action need coding, it's difficult for me because I haven't experience in coding with py gco.   sighed!

**Comment by Tai-Hsien** [comment]
The problem comes from the transfer between the downsampled resolution and original resolution. I use SVM or KNN to do the resolution transfer, which make the boundary blurry. To solve this problem, we could perform the fine-tune algorithm (e.g., graph cut) again under the original resolution. However, the process is quite slow because the high resolution contains too many faces (triangles) so I didn't include it in my scripts. If you plan to do it, you can take a look at step6 with pygco. I suggest using pygco instead of gco since pygco is more memory-efficient.

Alternatively, we could increase the resolution of downsampled mesh to reduce the difference between downsampled and original resolutions. However, this way also increase the use of memory, and not all computer can handle it.

Also, we can find a better way to transfer the segmentation result from downsampled to origin resolutions.

Lastly, if you just want to smooth boundary, you can save each tooth segmentation to a STL or OBJ file (I believe this is what you do) and use MeshMixer function to smooth the boundary.

Hope one of these ways will work for you.

**Comment by ly000451** [comment]
![image](https://user-images.githubusercontent.com/522072/103256338-df621a00-49c7-11eb-83e0-697e037abb99.png)
the upper image displays my MeshSegNet step 6 changed version.  Upsampling method using SVM will lead run error. so I change to KNN method. 
Below image is my segmented result.
![image](https://user-images.githubusercontent.com/522072/103256622-e89fb680-49c8-11eb-8bb2-2268a5dc4e48.png)

So you can see the boundary of the segmented tooth is no so accurate. It needs to adjust manually.
You can use Meshmixer to do this work, but Meshmixer can't recognize the labeled mesh. can't use MeshSegNet step 6 result. 
So I am in a dilemma.

**Comment by Tai-Hsien** [comment]
Hello @ly000451 

I see what you mean, and I just realize that your mesh is a **lower arch**, which is the major reason to cause this problem. The pre-trained model in my github is for upper only. Based on my previous experience, it works poorly for lower.

**Comment by ly000451** [comment]
I got it. Do you have a plan to optimize the pre-trained model and launched the special model for a lower arch tooth.  I am looking forward to it.  good job.

**Comment by Tai-Hsien** [comment]
I just uploaded a pre-trained model for lower arches. You can try it on your lower mesh to see if it works better.

**Comment by ly000451** [comment]
I have tried this pro-trained model. the result is Inspiring. Below image is my segmented mesh.
![image](https://user-images.githubusercontent.com/522072/103324557-f6b20d80-4a82-11eb-9a4e-bda92128a059.png)

Except for UR6, UR7,  the other tooth mesh is segmented perfectly.  Toward to UR6, UR7,   Is there any problem?

**Comment by Tai-Hsien** [comment]
No, there is no problem. It actually makes sense because the model was trained by a small sample size.
